KAFKA-14476: Move OffsetMap and related to storage module

ijuma opened the pull request:
For broader context on this change, please check:

* KAFKA-14470: Move log layer to storage module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma committed:
KAFKA-14476: Move OffsetMap and related to storage module

ijuma commented:
cc @satishd

satishd commented on storage/src/main/java/org/apache/kafka/server/log/internals/SkimpyOffsetMap.java:
```diff
@@ -0,0 +1,241 @@
+import java.util.Arrays;
+
+/**
+ * An hash table used for deduplicating the log. This hash table uses a cryptographicly secure hash of the key as a proxy for the key
```
minor existing typo in earlier code: "A" instead of "An"

ijuma replied:
Fixed.

satishd reviewed (commented):
Thanks @ijuma for the PR, left a minor comment.

satishd commented:
@ijuma There are test failures occurring in LogCleanerLagIntegrationTest, OffsetMapTest, and LogCleanerParameterizedIntegrationTest

satishd commented on storage/src/main/java/org/apache/kafka/server/log/internals/SkimpyOffsetMap.java:
```diff
@@ -0,0 +1,241 @@
+     */
+    @Override
+    public void put(ByteBuffer key, long offset) throws DigestException {
+        if (entries < slots)
```
It should be `entries >= slots`.
This condition was there in earlier scala code with require check. We need to invert it here.

ijuma replied:
Good catch, the check needs to be reversed since I switched Scala's from `require` to an `if`/`throw`. Fixed.

satishd reviewed (commented):
LogCleanerLagIntegrationTest, OffsetMapTest, and LogCleanerParameterizedIntegrationTest are failing because of this issue.

ijuma committed:
Address review comments

dengziming reviewed (approved):
LGTM.

ijuma committed:
Fix `bytesPerEntry` calculation

ijuma commented:
@satishd Addressed your comments and pushed an additional fix to address an `OffsetMapTest` failure. The CI build should be good now, but I'll wait for it to complete.

ijuma commented:
@dengziming Fyi, I pushed a small fix right after your review.

satishd reviewed (approved):
Thanks @ijuma for addressing review comments. LGTM.

fvaleri commented:
LGTM

ijuma committed:
Merge branch 'trunk' into kafka-14476-move-offset-map-to-storage

ijuma commented:
JDK 17 build passed and the other build failures are unrelated:

> Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testOffsetSyncsTopicsOnTarget()
> Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest.shouldPauseActiveTaskAndTransitToUpdateStandby()
> Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.internals.AbstractCoordinatorTest.testWakeupAfterSyncGroupReceivedExternalCompletion()
>
